File: Minimal/queuewithstacks.py
#!/usr/bin/python
import random
import stack


class QueueWithStacks():

    def __init__(self):
        ''' A queue that respects FIFO implemented with two stacks instead of
        a linked list. Stack B is responsible for dequeuing and Stack A for 
        enqueuing. '''
        self.stackA = stack.Stack()
        self.stackB = stack.Stack()
        self.size = 0

    def __repr__(self):
        _list = self.to_list()
        _list = map(str, _list)
        return ", ".join(_list)

    def to_list(self):
        invStack = self.stackA.inverted_copy()
        return self.stackB.to_list() + invStack.to_list()

    def is_empty(self):
        ''' Returns True if the queue is empty. False otherwise '''
        return self.size == 0

    def enqueue(self, value):
        ''' Insertions always occur in stackA '''
        self.stackA.push(value)
        self.size += 1

    def dequeue(self):
        ''' Removals always occur in StackB. If StackB is empty, all the
        content from StackA is moved to StackB '''
        if self.size == 0:
            return None
        else:
            if self.stackB.is_empty():
                while not self.stackA.is_empty():
                    element = self.stackA.pop()
                    self.stackB.push(element.value)

            element = self.stackB.pop()
            self.size -= 1
            return element


if __name__ == "__main__":
    q = QueueWithStacks()

    for i in range(10):
        q.enqueue(i)

    print q
    q.dequeue()
    q.enqueue(10)
    q.enqueue(11)
    print q

File: Algorithms/cpp/queue_with_stacks.cpp
#include <iostream>
#include <stack>
#include <queue>

using namespace std;

class queue_with_stacks
{
	stack<int> stack_in, stack_out;
public:
	queue_with_stacks();
	~queue_with_stacks();

	void enqueue(int);
	int dequeue();
};

queue_with_stacks::queue_with_stacks() {}
queue_with_stacks::~queue_with_stacks(){}

void queue_with_stacks::enqueue(int element)
{
	this->stack_in.push(element);
}
int queue_with_stacks::dequeue()
{
	int element;
	if (this->stack_out.empty())
	{
		while (!this->stack_in.empty())
		{
			this->stack_out.push(this->stack_in.top());
			this->stack_in.pop();
		}
	}
	if (!this->stack_out.empty())
	{
		element = this->stack_out.top();
		this->stack_out.pop();
		return element;
	}
	else
	{
		cout << "Queue is empty" << endl;
		return -1;
	}

}


int main(int argc, char ** argv)
{
	queue_with_stacks q;
	q.enqueue(1);
	q.enqueue(2);
	cout << q.dequeue() << endl;
	q.enqueue(3);
	q.enqueue(4);
	cout << q.dequeue() << endl;
	cout << q.dequeue() << endl;
	cout << q.dequeue() << endl;
	return 1;
}